src/reporter.cc

```diff
@@ -71,46 +78,109 @@ void MetadataReporter::ReportMetadata() {
 namespace {
 
 void SendMetadataRequest(std::vector<json::value>&& entries,
-                         const std::string& endpoint,
+                         const std::string& batch_uri,
+                         const std::string& publish_path,
                          const std::string& auth_header,
                          const std::string& user_agent,
                          bool verbose_logging)
     throw (boost::system::system_error) {
-  json::value update_metadata_request = json::object({
-    {"entries", json::array(std::move(entries))},
-  });
 
-  if (verbose_logging) {
-    LOG(INFO) << "About to send request: POST " << endpoint
-              << " User-Agent: " << user_agent
-              << " " << *update_metadata_request;
+  if (entries.size() == 1) {
+    // Add a copy of the single entry, except for the `views` field. This allows
+    // us to sent a request to the batch endpoint when we have a single request.
+    const json::Object* single_request = entries[0]->As<json::Object>();
+    json::value duplicate_request = json::object({  // ResourceMetadata
+      {"name", json::string(single_request->Get<json::String>("name"))},
+      {"type", json::string(single_request->Get<json::String>("type"))},
+      {"location", json::string(single_request->Get<json::String>("location"))},
+      {"state", json::string(single_request->Get<json::String>("state"))},
+      {"eventTime",
+        json::string(single_request->Get<json::String>("eventTime"))},
+    });
+    entries.emplace_back(std::move(duplicate_request));
   }
-
+  const std::string content_type =
+      std::string("multipart/mixed; boundary=") + kMultipartBoundary;
   http::client client;
-  http::client::request request(endpoint);
-  std::string request_body = update_metadata_request->ToString();
+  http::client::request request(batch_uri);
   request << boost::network::header("User-Agent", user_agent);
-  request << boost::network::header("Content-Length",
-                                    std::to_string(request_body.size()));
-  request << boost::network::header("Content-Type", "application/json");
+  request << boost::network::header("Content-Type", content_type);
   request << boost::network::header("Authorization", auth_header);
-  request << boost::network::body(request_body);
+  request << boost::network::header("Expect", "100-continue");
+  std::ostringstream out;
+  out << std::endl;
+
+  for (json::value& entry : entries) {
+    const json::Object* single_request = entry->As<json::Object>();
+    std::string request_body = single_request->ToString();
+    out << "--" << kMultipartBoundary << std::endl;
```

Comment:
This might look more readable as a multi-line format string, e.g.:
```c++
constexpr const char kPartFormat[] =
    "-- {{boundary}}\n"
    "Content-Type: application/http\n"
    "Content-Transfer-Encoding: binary\n"
    "Content-ID: {{name}}\n"
    "\n"
    "POST {{path}}\n"
    "Content-Type: application/json; charset=UTF-8\n"
    "ContentLength: {{length}}\n"
    "\n"
    "{{body}}\n";
...
  for (json::value& entry : entries) {
    const json::Object* single_request = entry->As<json::Object>();
    std::string request_body = single_request->ToString();
    out << format::Substitute(kPartFormat, {
      {"boundary", kMultipartBoundary},
      {"name", single_request->Get<json::String>("name")},
      {"path", publish_path},
      {"length", std::to_string(request_body.size())},
      {"body", request_body},
    });
  }
```